### PR TITLE
ET-4598 string filter

### PR DESCRIPTION
**Pull request by janpetschexain**
**Reference**

- [ET-4598]
- followed by #1003

**Summary**

- declare string filters in the spec & update changelog
- impl compare filter serde & add them in personalization routes
- add compare filter in elasticsearch calls

[ET-4598]: https://xainag.atlassian.net/browse/ET-4598?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: web-api/tests/document_candidates.rs [hunk 1]
```diff
@@ -71,65 +77,81 @@ async fn set(client: &Client, url: &Url, ids: impl IntoIterator<Item = &str>) ->
...
 fn test_candidates_all() {
-    test_app::<Ingestion, _>(UNCHANGED_CONFIG, |client, url, _| async move {
-        assert!(get(&client, &url).await?.ids().is_empty());
-        ingest(&client, &url).await?;
-        assert_eq!(get(&client, &url).await?.ids(), ["d1", "d2", "d3"].into());
-        set(&client, &url, ["d1", "d2", "d3"]).await?;
-        assert_eq!(get(&client, &url).await?.ids(), ["d1", "d2", "d3"].into());
-        Ok(())
-    });
+    test_app::<Ingestion, _>(
```
Comment (janpetschexain) [inline web-api/tests/document_candidates.rs:80]: all the changes in the integration tests, except for the new test for the filter, are just indention formatting changes due to the new argument in `test_app()`.

File: web-api/src/personalization/filter.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                            ));
+                        }
+                    }
+                }
+
+                Ok(CompareWith { operation, value })
+            }
+        }
+
+        deserializer.deserialize_map(CompareWithVisitor)
```
Comment (rustonaut) [inline web-api/src/personalization/filter.rs:80]: :question: I'm not sure if I'm missing something but why do it so complicated.

Couldn't we have used something like:

```rust
#[derive(Deserialize)]
struct LeafFilter {
    property: DocumentPropertyId,
    operator: LeafFilterOperator,
}

#[derive(Serialize)]
enum LeafFilterOperator {
  #[serde(rename="$eq")]
  Eq(AnyPropertyValue),
  //... Ne, Lt, Lte, In, etc. 
  #[serde(rename="$lt")]
  Lt(OrderedPropertyValue),
  #[serde(rename="$in")]
  In(Vec<DocumentPropertyString>),
}
```

or if you want to have better handcrafted error messages but in turn also more 
post evaluation by hand:

```rust
type LeafFilterOperator = HashMap<LeafOperator, Value>; 
``` 

sure for `$and`/`$or` this will still require some usage of `#[serde(untagged)]` or even
manual code, but it should be in general easier to implement this way.

File: web-api/src/storage/elastic.rs [hunk 3]
```diff
@@ -407,6 +408,27 @@ impl KnnSearchParams<'_> {
                 json!({ "range": { "properties.publication_date": { "gte": published_after } } }),
             );
         }
+        if let Some(opt_filter) = self.filter {
+            let opt_filter = match opt_filter {
+                Filter::Compare(compare) => match compare.operation {
+                    CompareOp::Eq => {
+                        json!({ "term": { format!("properties.{}", compare.field): compare.value } })
```
Comment (rustonaut) [inline web-api/src/storage/elastic.rs:415]: This probably works but elasticsearch documents it as:

```json
{ "term": { "property.name": { "value": "comparision_value" } } }
```

https://www.elastic.co/guide/en/elasticsearch/reference/8.8/query-dsl-term-query.html

funny thing is with `terms` it does not do so :face_exhaling: 

```suggestion
                        json!({ "term": { format!("properties.{}", compare.field): { "value":  compare.value } } })
```

**Review by rustonaut — COMMENTED** [review]
:pushpin:  We need to add checks which compare the current schema to the operators.

E.g. something walk over all leaf operators and check if the property for them is in the schema and if if the type in the schema and operator matches (note that for Eq/Ne `StringArray in schema matches String in query` and is the equivalent of "is one of the properties values (not)equal to given comparison string", this is in line with how both elasticsearch and pinecone works).

Not doing so allows some edge cases in which the user can run queries which we can not guarantee to support in the future.

But that definitely has less priority then supporting the `$in` operator.

File: integration-tests/src/lib.rs [hunk 4]
```diff
@@ -478,6 +478,7 @@ impl Drop for DeleteTempDirIfNoPanic {
 /// - the service info including an url to the application is passed to the test
 pub fn test_app<A, F>(
     configure: Option<Table>,
+    indexable_properties: Option<&'static [(&'static str, &'static str)]>,
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply integration-tests/src/lib.rs:481]: Just saw your comment in my PR :sweat_smile:

File: web-api/openapi/front_office.yaml [hunk 5]
```diff
@@ -152,6 +159,21 @@ components:
...
       default: true
+    FilterString:
+      type: object
+      additionalProperties:
+        type: object
+        properties:
+          $eq:
+            $ref: ./schemas/document.yml#/DocumentPropertyString
+        minProperties: 1
+        maxProperties: 1
```
[the remark this answers is not in the document]
Comment (janpetschexain) [reply web-api/openapi/front_office.yaml:170]: with `FilterEq` you skip one level of nesting. basically we want compare filter queries of the form
```
{ <property_name>: { <compare_op>: <property_value>    } }
|<FilterString>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>|
                   |<FilterString.additionalProperties>|
```

File: web-api/src/personalization/filter.rs (hunk 2 above)
Comment (janpetschexain) [reply web-api/src/personalization/filter.rs:80]: this was pretty much my line of thought, i went from enum to hashmap to singlevaluehashmap. the main difference is that i flattened the `SingleValueMap<DocumentPropertyId, SingleValueMap<LeafOperator, Value>>` which is essentially the `Compare` and the inner `SingleValueMap<LeafOperator, Value>` is the `CompareWith` and the `LeafOperator` is the `CompareOp`.
with `Compare` and `CompareWith` we can get better error message where the input was invalid, with just a generic `SingleValueMap` this would be more difficult. if that's not needed, than i can also make it more generic, but it seemed good to have.

File: integration-tests/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (janpetschexain) [reply integration-tests/src/lib.rs:481]: so i guess il rebase once #997 is merged to have the least merge conflicts

File: web-api/openapi/front_office.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/openapi/front_office.yaml]: What I'm saying is:

```
{ <property_id>: { <compare_op>: <property_value>             } }
|<LeafFilter object with additionalProperties>>>>>>>>>>>>>>>>>>>|
                 |<FilterEq object>>>>>>>>>>>>>>>>>>>>>>>>>>>|
```

[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/openapi/front_office.yaml]: Because we don't only have string filter and the object  which contains the property id shouldn't be specific to any type of the filter operation it contains.

[the remark this answers is not in the document]
Comment (janpetschexain) [reply web-api/openapi/front_office.yaml]: if i'm not mistaken this is already facilitated by renaming the `FilterString` to `FilterCompare` (currently its inner object only has one property, but that's equivalent to `oneOf` with a single item anyways). so for example the current definition is
```yaml
FilterCompare:
  type: object
  additionalProperties:
    type: object
    properties:
      $eq:
        $ref: ./schemas/document.yml#/DocumentPropertyString
    minProperties: 1
    maxProperties: 1
```
and in the following prs it will be extended to for example
```yaml
FilterCompare:
  type: object
  additionalProperties:
    type: object
    properties:
      $eq:
        oneOf:
          - $ref: ./schemas/document.yml#/DocumentPropertyString
          - $ref: ./schemas/document.yml#/DocumentPropertyBool
      $in:
        $ref: ./schemas/document.yml#/DocumentPropertyArrayString
    minProperties: 1
    maxProperties: 1
```
this is essentially the same, just without naming the inner object.

File: integration-tests/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (janpetschexain) [reply integration-tests/src/lib.rs]: rebasing is done

**Review by rustonaut — APPROVED** [review]
